Add the layers drop_path.py, layer_scale.py, mlp.py, swiglu_ffn.py of DINOV2

mohamedamrfaroukebrahim opened the pull request.

mohamedamrfaroukebrahim committed:
Add drop_path.py, layer_scale.py, mlp.py, swiglu_ffn.py layers of DINOV2

mohamedamrfaroukebrahim committed:
fix issues in drop_path.py and swinglu_ffn.py

oarriaga commented on paz/models/foundation/dinov2/drop_path.py:
```diff
@@ -0,0 +1,42 @@
+from typing import Optional
+
+
+def drop_path_keras_functional(
```
We don't need to write "keras"

oarriaga commented on paz/models/foundation/dinov2/drop_path.py:
```diff
@@ -0,0 +1,42 @@
+
+
+def drop_path_keras_functional(
+    x: keras.KerasTensor, drop_prob: float = 0.0, training: bool = False
```
We are not using type annotations.

oarriaga replied:
dropout_rate

oarriaga commented on paz/models/foundation/dinov2/drop_path.py:
```diff
@@ -0,0 +1,42 @@
+    # random_tensor = ops.random.bernoulli(shape_for_random, p=keep_prob, dtype=x.dtype)
+    random_tensor = keras.random.binomial(
+        shape=shape_for_random, counts=1.0, probabilities=keep_prob, dtype=x.dtype
+    )
```
Wrap this as a function that just does bernoulli.

oarriaga commented on paz/models/foundation/dinov2/mlp.py:
```diff
@@ -0,0 +1,30 @@
+from typing import Optional, Type
+
+
+class Mlp(keras.Model):
```
Acronyms should always be capitalized.

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Fix Requested changes

mohamedamrfaroukebrahim committed:
add dino_head.py to DINOv2 foundation model

mohamedamrfaroukebrahim committed:
remove all type annotations

mohamedamrfaroukebrahim committed:
add patch_embed.py to DINOv2 foundation model

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Add `block.py` to dinov2 foundation model

mohamedamrfaroukebrahim committed:
fix `dino_head.py`

mohamedamrfaroukebrahim committed:
Add DINO model components including layers and vision transformer architecture

- Implement WeightNormDense and DINOHead layers for the DINO model.
- Create DropPath, LayerScale, MLP, PatchEmbed, and SwiGLUFFN layers for enhanced functionality.
- Introduce vision transformer model with various configurations (small, base, large, giant2).
- Add utility functions for model building and weight initialization.
- Ensure compatibility with Keras and maintain modular structure for easy adjustments.

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Remove deprecated DINO model components and clean up related files

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Add DINOv2 model testing and weight porting functionality

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+        dimension,
+        num_heads=8,
+        qkv_bias=False,
+        projection_bias=True,
```
Let's indicate that they are boolean.

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+        self.projection_bias = projection_bias
+        self.attention_drop_rate_val = attention_drop_rate
+        self.projection_drop_rate_val = projection_drop_rate
+        self.head_dim = self.dimension // self.num_heads
```
Let's always use full variable names "dim" for "dimension"

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+        self.qkv_bias = qkv_bias
+        self.projection_bias = projection_bias
+        self.attention_drop_rate_val = attention_drop_rate
+        self.projection_drop_rate_val = projection_drop_rate
```
Let's keep the same variable name and not change it internally in the class.

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+
+        initializer = keras.initializers.TruncatedNormal(stddev=0.02)
+        self.qkv = Dense(
+            self.dimension * 3, use_bias=self.qkv_bias, kernel_initializer=initializer, name="qkv"
```
Maybe to make it fit in one line, we could do kwargs = {"kernel_initializer": initializer} and pass the kwargs to both dense functions.

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+        self.projection_drop = Dropout(self.projection_drop_rate_val)
+
+    def build(self, input_shape):
+        super().build(input_shape)
```
Let's try to remove it, since we are not doing anything with it.

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+        super().build(input_shape)
+
+    def call(self, x, attn_bias=None, training=None):
+        B, N, C = ops.shape(x)
```
Let's rename this to full variable names. batch_size or batch, ...

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+        attn = ops.matmul(q, ops.transpose(k, axes=[0, 1, 3, 2])) * self.scale
+
+        if attn_bias is not None:
+            attn += attn_bias
```
Let's be more explicit: attention = attention + attention_bias

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+
+    def call(self, x, attn_bias=None, training=None):
+        B, N, C = ops.shape(x)
+        qkv = self.qkv(x)
```
Maybe we should be able to read this as "self.qkv" -> "self.predict_query_key_value"

oarriaga commented on paz/models/foundation/dinov2/layers/attention.py:
```diff
@@ -0,0 +1,65 @@
+
+class MemEffAttention(Attention):
+    def call(self, x, attn_bias=None, training=None):
+        return super().call(x, attn_bias=attn_bias, training=training)
```
We should try to remove this.

oarriaga commented on paz/models/foundation/dinov2/layers/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from .swiglu_ffn import SwiGLUFFN, SwiGLUFFNFused, SwiGLUFFNAligned
+from .block import NestedTensorBlock, CausalAttentionBlock
+from .attention import Attention, MemEffAttention
+from .drop_path import DropPath
```
Let's do the imports always from root:
"from paz.models.foundation.layers import DINOHead"

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+    sample_subset_size = ops.maximum(ops.cast(subset_float, "int32"), 1)
+
+    all_indices = ops.arange(b, dtype="int32")
+    brange = keras.random.shuffle(all_indices)[:sample_subset_size]
```
Rename to something that conveys more meaning to the reader.

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+):
+    b = ops.shape(x)[0]
+
+    subset_float = ops.multiply(ops.cast(b, "float32"), (1.0 - sample_drop_ratio))
```
why is this a subset

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+
+    x_subset = ops.take(x, brange, axis=0)
+
+    residual = residual_func(x_subset)
```
apply_residual

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+        scatter_indices = ops.expand_dims(brange, axis=1)
+        x_plus_residual_flat = ops.scatter_update(x_flat, scatter_indices, updated_values)
+
+        return ops.reshape(x_plus_residual_flat, original_shape)
```
let's just have a single return

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+                block = ops.zeros((rows, cols), dtype=dtype)
+            block_row.append(block)
+        rows_of_blocks.append(ops.concatenate(block_row, axis=1))
+    return ops.concatenate(rows_of_blocks, axis=0)
```
Let's try to understand what this function is doing and change the name to show me that.

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+    return ops.concatenate(rows_of_blocks, axis=0)
+
+
+def get_attn_bias_and_cat(x_list, branges=None):
```
same as before let's try to understand this.

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+        self.feed_forward.fc2.kernel.assign(
+            keras.random.normal(self.feed_forward.fc2.kernel.shape, stddev=init_proj_std)
+        )
+
```
Let's try to encapsulate this into a function that tells me what is happening. We can then try to use other functions such as initializers?

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+                sample_drop_ratio=self.sample_drop_ratio,
+                scaling_vector=self.ls2.gamma if isinstance(self.ls2, LayerScale) else None,
+            )
+            return x_list
```
one return

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+        else:
+
+            def attn_residual_func(x, attn_bias=None):
+                return self.ls1(self.attn(self.norm1(x), attn_bias=attn_bias))
```
I don't know what ls1 is. Let's use full variable names of attention, normalization, etc.

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+            def ffn_residual_func(x, attn_bias=None):
+                return self.ls2(self.mlp(self.norm2(x)))
+
+            attn_mask, x_cat = get_attn_bias_and_cat(x_list)
```
let's use full variable names. Let's not use cat.

oarriaga commented on paz/models/foundation/dinov2/layers/block.py:
```diff
@@ -0,0 +1,379 @@
+        elif hasattr(x_or_x_list, "shape") and hasattr(x_or_x_list, "dtype"):
+            return super().call(x_or_x_list, training=training)
+        else:
+            raise TypeError(f"Input must be a tensor or a list of tensors, but got {type(x_or_x_list)}")
```
Let's double check that we really need this logic here.

oarriaga commented on paz/models/foundation/dinov2/layers/dino_head.py:
```diff
@@ -0,0 +1,136 @@
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+
+        self.v = self.add_weight(
```
what is v?

oarriaga commented on paz/models/foundation/dinov2/layers/dino_head.py:
```diff
@@ -0,0 +1,136 @@
+            name="v",
+        )
+
+        self.g = self.add_weight(shape=(self.units,), initializer="ones", trainable=True, name="g")
```
what is g?

oarriaga commented on paz/models/foundation/dinov2/layers/dino_head.py:
```diff
@@ -0,0 +1,136 @@
+
+        x = ops.divide(x, ops.maximum(ops.sqrt(ops.sum(ops.square(x), axis=-1, keepdims=True)), eps))
+
+        x = self.last_layer(x)
```
Maybe this name can be different? It's clear that is the last layer. But maybe is not clear that is weight normalization dense?

oarriaga commented on paz/models/foundation/dinov2/layers/dino_head.py:
```diff
@@ -0,0 +1,136 @@
+        return x
+
+
+def _build_mlp(nlayers, in_dim, bottleneck_dim, hidden_dim=None, use_bn=False, bias=True):
```
complicated logic to read right away, let's try to make simpler to read by either making functions and full variable names.

oarriaga commented on paz/models/foundation/dinov2/layers/layer_scale.py:
```diff
@@ -0,0 +1,31 @@
+class LayerScale(keras.layers.Layer):
+    def __init__(self, dim, init_values=1e-5, inplace=False, dtype=None, **kwargs):
+        super().__init__(dtype=dtype, **kwargs)
+        self.dim = dim
```
let's use full variable names dimension

oarriaga commented on paz/models/foundation/dinov2/layers/mlp.py:
```diff
@@ -0,0 +1,33 @@
+        x = self.drop_layer(x, training=training)
+        x = self.fc2(x)
+        x = self.drop_layer(x, training=training)
+        return x
```
I think we could replace this too be something like :

def block_MLP(x, some_params):
     x = Dense(params)(x)
     x = Activation(params)(x)
     ....

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Apply requested changes those did not affect the test cases

mohamedamrfaroukebrahim committed:
Apply the rest of the requested changes and added some changes for readability

mohamedamrfaroukebrahim committed:
Refactor and enhance DINOv2 model layers and weight porting

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Merge branch 'paz-jax' of https://github.com/mohamedamrfaroukebrahim/paz into paz-jax

mohamedamrfaroukebrahim committed:
Remove unused FLOPS calculation method from PatchEmbed class

mohamedamrfaroukebrahim committed:
Add Dinov2_giant model building, weights porting, testing and potential issues fixing

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax

mohamedamrfaroukebrahim committed:
Merge branch 'paz-jax' of https://github.com/mohamedamrfaroukebrahim/paz into paz-jax

mohamedamrfaroukebrahim committed:
Merge branch 'oarriaga:paz-jax' into paz-jax
